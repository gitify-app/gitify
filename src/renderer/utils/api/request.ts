import axios, {
  type AxiosResponse,
  type AxiosPromise,
  type Method,
} from 'axios';
import log from 'electron-log';
import type { Link, Token } from '../../types';
import { getNextURLFromLinkHeader } from './utils';

/**
 * Perform an unauthenticated API request
 *
 * @param url
 * @param method
 * @param data
 * @returns
 */
export function apiRequest(
  url: Link,
  method: Method,
  data = {},
): AxiosPromise | null {
  axios.defaults.headers.common.Accept = 'application/json';
  axios.defaults.headers.common['Content-Type'] = 'application/json';
  axios.defaults.headers.common['Cache-Control'] = 'no-cache';
  return axios({ method, url, data });
}

/**
 * Perform an authenticated API request
 *
 * @param url
 * @param method
 * @param token
 * @param data
 * @param fetchAllRecords whether to fetch all records or just the first page
 * @returns
 */
export async function apiRequestAuth(
  url: Link,
  method: Method,
  token: Token,
  data = {},
  fetchAllRecords = false,
): AxiosPromise | null {
  axios.defaults.headers.common.Accept = 'application/json';
  axios.defaults.headers.common.Authorization = `token ${token}`;
  axios.defaults.headers.common['Content-Type'] = 'application/json';
  axios.defaults.headers.common['Cache-Control'] = shouldRequestWithNoCache(url)
    ? 'no-cache'
    : '';

  if (!fetchAllRecords) {
    return axios({ method, url, data });
  }

  let response: AxiosResponse | null = null;
  let combinedData = [];

  try {
    let nextUrl: string | null = url;

    while (nextUrl) {
      response = await axios({ method, url: nextUrl, data });

      // If no data is returned, break the loop
      if (!response?.data) {
        break;
      }

      combinedData = combinedData.concat(response.data); // Accumulate data

      nextUrl = getNextURLFromLinkHeader(response);
    }
  } catch (error) {
    log.error('API request failed:', error);
    throw error;
  }

  return {
    ...response,
    data: combinedData,
  } as AxiosResponse;
}

/**
 * Return true if the request should be made with no-cache
 *
 * @param url
 * @returns boolean
 */
function shouldRequestWithNoCache(url: string) {
  const parsedUrl = new URL(url);

  switch (parsedUrl.pathname) {
    case '/notifications':
      return true;
    default:
      return false;
  }
}
