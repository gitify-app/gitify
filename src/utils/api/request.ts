import axios, { type AxiosPromise, type Method } from 'axios';
import type { Link, Token } from '../../types';

export function apiRequest(
  url: Link,
  method: Method,
  data = {},
): AxiosPromise | null {
  axios.defaults.headers.common.Accept = 'application/json';
  axios.defaults.headers.common['Content-Type'] = 'application/json';
  axios.defaults.headers.common['Cache-Control'] = 'no-cache';
  return axios({ method, url, data });
}

export function apiRequestAuth(
  url: Link,
  method: Method,
  token: Token,
  data = {},
): AxiosPromise | null {
  axios.defaults.headers.common.Accept = 'application/json';
  axios.defaults.headers.common.Authorization = `token ${token}`;
  axios.defaults.headers.common['Cache-Control'] = 'no-cache';
  axios.defaults.headers.common['Content-Type'] = 'application/json';
  return axios({ method, url, data });
}
