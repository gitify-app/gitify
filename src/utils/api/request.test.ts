import axios from 'axios';
import type { Link, Token } from '../../types';
import { apiRequest, apiRequestAuth } from './request';

jest.mock('axios');

const url = 'https://example.com' as Link;
const method = 'get';

describe('utils/api/request.ts', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should make a request with the correct parameters', async () => {
    const data = { key: 'value' };

    await apiRequest(url, method, data);

    expect(axios).toHaveBeenCalledWith({
      method,
      url,
      data,
    });

    expect(axios.defaults.headers.common).toMatchSnapshot();
  });

  it('should make a request with the correct parameters and default data', async () => {
    const data = {};
    await apiRequest(url, method);

    expect(axios).toHaveBeenCalledWith({
      method,
      url,
      data,
    });

    expect(axios.defaults.headers.common).toMatchSnapshot();
  });
});

describe('apiRequestAuth', () => {
  const token = 'yourAuthToken' as Token;

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should make an authenticated request with the correct parameters', async () => {
    const data = { key: 'value' };

    await apiRequestAuth(url, method, token, data);

    expect(axios).toHaveBeenCalledWith({
      method,
      url,
      data,
    });

    expect(axios.defaults.headers.common).toMatchSnapshot();
  });

  it('should make an authenticated request with the correct parameters and default data', async () => {
    const data = {};

    await apiRequestAuth(url, method, token);

    expect(axios).toHaveBeenCalledWith({
      method,
      url,
      data,
    });

    expect(axios.defaults.headers.common).toMatchSnapshot();
  });
});
