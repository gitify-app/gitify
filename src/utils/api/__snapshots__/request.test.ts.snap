// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiRequestAuth should make an authenticated request with the correct parameters 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`apiRequestAuth should make an authenticated request with the correct parameters and default data 1`] = `
{
  "Accept": "application/json",
  "Authorization": "token yourAuthToken",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/request.ts should make a request with the correct parameters 1`] = `
{
  "Accept": "application/json",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;

exports[`utils/api/request.ts should make a request with the correct parameters and default data 1`] = `
{
  "Accept": "application/json",
  "Cache-Control": "no-cache",
  "Content-Type": "application/json",
}
`;
